src/ripple/core/impl/SociDB.cpp

```diff
@@ -243,15 +233,21 @@ class WALCheckpointer : public Checkpointer
         }
 
         // If the Job is not added to the JobQueue then we're not running_.
-        if (!jobQueue_.addJob(jtWAL, "WAL", [this](Job&) { checkpoint(); }))
+        if (!jobQueue_.addJob(
+                jtWAL,
+                "WAL",
+                [wp = std::weak_ptr<Checkpointer>{shared_from_this()}](Job&) {
```

Comment:
Why do we capture the `weak_ptr` instead of `shared_ptr`? The way this is written, if the Checkpointer was deleted before the job is run, the job is a noop. When does this occur? Maybe add a comment talking about this.

Reply:
This occurs when the `DatabaseCon` object has been destroyed. When this happens, the `conn_` member on the `Checkpointer` is no longer valid, so we can't call the checkpointer.

However, while I was writing this, I realized this patch still has a memory bug (tho much less likely than before). If the weak pointer is locked in the callback and then the database object is destoryed we'll access a bad connection. Will fix.